unison-cli/src/Unison/Codebase/Editor/HandleInput/Update.hs

```diff
@@ -189,12 +201,324 @@ handleUpdate input optionalPatch requestedNames = do
+    --
+    -- The running example here will be one in which the current namespace already has the following terms stored.
+    -- (For simplicity, we will ignore the limitation that mutually recursive definitions must be top-level arrows;
+    -- pretend this is Haskell's let).
+    --
+    --   ping = pong + 1   (reference = #pingpong.ping)
+    --   pong = ping + 2   (reference = #pingpong.pong)
+    --   wham = pong + 3   (reference = #wham)
+    --
+    -- The user re-defines "ping" in their scratch file thus:
+    --
+    --   ping = wham + 4   (reference = #newping)
+    --
+    -- Note that, pre-update, we had two components [ping,pong] and [wham]. But after the update, since the new `ping`
+    -- refers to the old `wham`, which refers to the old `pong`, which refers to the old `ping`, we really want to end
+    -- up with a single component in the end, [ping,pong,wham].
+
+    -- First, compute an initial slurp, which will identify the initial set of definitions we are updating ({"ping"}).
+    -- Also, this will be the slurp that we fall back on, in case re-typechecking another Unison file with implicit
+    -- terms in it fails.
+    let slurp0 = slurp unisonFile0
+
+    -- Get the set of names that are being updated.
+    --
+    -- Running example:
+    --
+    --   { "ping" }
+    let namesBeingUpdated :: Set Symbol
+        namesBeingUpdated =
+          SC.terms (Slurp.updates slurp0)
+
+    -- Associate each such name with the set of "old" references (already in the codebase) that it's associated with.
+    --
+    -- Running example:
+    --
+    --   "ping" => { #pingpong.ping }
+    let updatedNameToOldRefs :: Map Symbol (Set TermReference)
+        updatedNameToOldRefs =
+          Map.fromSet nameToTermRefs namesBeingUpdated
+
+    -- Identify all of the implicit terms, which are:
+    --
+    --   - Both:
+    --     - Either:
+    --         - (A) A component-mate of a term that's being updated.
+    --         - Both:
+    --           - (B) A dependent of a term that's being updated.
+    --           - (C) A dependency of the new term.
+    --     - (D) Not being updated.
+    --
+    -- FIXME Additionally, we have two more temporary requirements.
+    --
+    --   - (E) The term has exactly one name in the namespace.
+    --   - (F) No other terms share its name.
+    --
+    -- This works around two fixable issues:
+    --
+    --   1. If the term has no names in the namespace, then we can't successfully put it into the Unison file anyway and
+    --      forward it through the typecheck + slurp process, because slurping currently assumes that it can freely
+    --      convert back-and-forth between vars and names.
+    --
+    --   2. If more than one term has its name, then putting this name in the Unison file and proceeding to do an update
+    --      would have the undesirable side-effect of resolving the conflict.
+    --
+    -- FIXME don't bother for type-changing updates
```

Comment:
FIXME

Reply:
Also being tracked in the minimegaticket